File: AssetSync/src/AssetSync.Infrastructure/Api/SnipeItService.cs
using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using AssetSync.Core.Interfaces;
using AssetSync.Core.Models;

namespace AssetSync.Infrastructure.Api;

public class SnipeItService : ISnipeItService
{
    private const int MaxRetries = 3;
    private readonly string _baseUrl;
    private readonly Func<string> _getApiKey;
    private readonly IHttpClientFactory _httpClientFactory;

    public SnipeItService(string baseUrl, Func<string> getApiKey, IHttpClientFactory httpClientFactory)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _getApiKey = getApiKey;
        _httpClientFactory = httpClientFactory;
    }

    public async Task<Device?> GetAssetBySerialAsync(string normalizedSerial, CancellationToken cancellationToken = default)
    {
        var list = await SearchAssetsBySerialAsync(normalizedSerial, cancellationToken).ConfigureAwait(false);
        return list.Count == 1 ? list[0] : null;
    }

    public async Task<IReadOnlyList<Device>> SearchAssetsBySerialAsync(string searchTerm, CancellationToken cancellationToken = default)
    {
        // Use the dedicated byserial endpoint for exact matching — avoids broad search misses
        var url = $"{_baseUrl}/api/v1/hardware/byserial/{Uri.EscapeDataString(searchTerm)}";
        var response = await SendWithRetryAsync(HttpMethod.Get, url, null, cancellationToken).ConfigureAwait(false);

        // 404 means no asset found — not an error
        if (response.StatusCode == System.Net.HttpStatusCode.NotFound)
            return new List<Device>();

        if (!response.IsSuccessStatusCode)
            return new List<Device>();

        var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;

        // byserial can return a single object or a {"rows":[...]} collection
        JsonElement rows;
        if (root.ValueKind == JsonValueKind.Array)
            rows = root;
        else if (root.ValueKind == JsonValueKind.Object && root.TryGetProperty("rows", out var rowsEl) && rowsEl.ValueKind == JsonValueKind.Array)
            rows = rowsEl;
        else if (root.ValueKind == JsonValueKind.Object && root.TryGetProperty("id", out _))
        {
            // Single asset returned directly
            var single = ParseDevice(root);
            return single != null ? new List<Device> { single } : new List<Device>();
        }
        else
            return new List<Device>();

        var list = new List<Device>();
        if (rows.ValueKind != JsonValueKind.Array) return list;
        foreach (var r in rows.EnumerateArray())
        {
            if (r.ValueKind != JsonValueKind.Object) continue;
            var device = ParseDevice(r);
            if (device != null) list.Add(device);
        }
        return list;
    }

    private static Device? ParseDevice(JsonElement r)
    {
        if (r.ValueKind != JsonValueKind.Object) return null;
        var id = r.TryGetProperty("id", out var idEl) ? idEl.GetInt32() : 0;
        var serial = r.TryGetProperty("serial", out var s) ? s.GetString() : null;
        // Snipe-IT HTML-encodes special characters (e.g. apostrophes become &#039;) — decode before storing
        var name = r.TryGetProperty("name", out var n) ? WebUtility.HtmlDecode(n.GetString()) : null;
        var modelId = r.TryGetProperty("model", out var mo) && mo.ValueKind == JsonValueKind.Object && mo.TryGetProperty("id", out var mid) ? mid.GetInt32() : (int?)null;
        var assignedTo = r.TryGetProperty("assigned_to", out var at) && at.ValueKind == JsonValueKind.Object && at.TryGetProperty("id", out var aid) ? aid.GetInt32() : (int?)null;
        var assetTag = r.TryGetProperty("asset_tag", out var tag) ? WebUtility.HtmlDecode(tag.GetString()) : null;
        return new Device
        {
            SnipeItAssetId = id,
            SerialNumber = serial,
            NormalizedSerial = serial != null ? Core.Services.SerialNumberNormalizer.Normalize(serial) : "",
            DeviceName = name,
            SnipeItModelId = modelId,
            SnipeItAssignedUserId = assignedTo,
            SnipeItAssetTag = assetTag
        };
    }

    public async Task<Device?> CreateAssetAsync(Device device, CancellationToken cancellationToken = default)
    {
        var payload = new Dictionary<string, object?>
        {
            ["name"] = device.DeviceName ?? device.SerialNumber,
            ["serial"] = device.SerialNumber,
            ["model_id"] = device.SnipeItModelId,
            ["status_id"] = 1, // Default to "Ready to Deploy"
        };
        // Only include optional fields when they have values — omitting null avoids Snipe-IT validation errors
        if (device.SnipeItAssignedUserId.HasValue)
            payload["assigned_to"] = device.SnipeItAssignedUserId.Value;
        if (device.SnipeItCategoryId.HasValue)
            payload["category_id"] = device.SnipeItCategoryId.Value;
        // Use the MDM asset tag if present, otherwise generate a unique TEMP tag (e.g. TEMP1A2B3C4D)
        // Snipe-IT requires an asset_tag on create. Use GUID-derived hex for ~4B combinations.
        var assetTag = !string.IsNullOrEmpty(device.MdmAssetTag)
            ? device.MdmAssetTag
            : $"TEMP{Guid.NewGuid().ToString("N")[..8].ToUpperInvariant()}";
        payload["asset_tag"] = assetTag;

        var payloadJson = JsonSerializer.Serialize(payload);
        var response = await SendWithRetryAsync(HttpMethod.Post, $"{_baseUrl}/api/v1/hardware", () => new StringContent(payloadJson, Encoding.UTF8, "application/json"), cancellationToken).ConfigureAwait(false);
        var responseJson = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);

        // Snipe-IT returns HTTP 200 even on validation errors — check the status field in the body
        JsonDocument doc;
        try { doc = JsonDocument.Parse(responseJson); }
        catch { return null; }

        var root = doc.RootElement;

        // {"status":"error","messages":{...}} means the create failed
        if (root.ValueKind == JsonValueKind.Object &&
            root.TryGetProperty("status", out var statusEl) &&
            statusEl.GetString() == "error")
        {
            // Surface the actual validation message so callers can log it
            var messages = root.TryGetProperty("messages", out var msgEl) ? msgEl.GetRawText() : responseJson;
            throw new InvalidOperationException($"Snipe-IT create rejected: {messages}");
        }

        if (!response.IsSuccessStatusCode) return null;

        var created = root.ValueKind == JsonValueKind.Object && root.TryGetProperty("payload", out var p) ? p : root;
        var id = created.ValueKind == JsonValueKind.Object && created.TryGetProperty("id", out var idEl) ? idEl.GetInt32() : 0;
        device.SnipeItAssetId = id;
        return device;
    }

    public async Task<bool> UpdateAssetAsync(int assetId, IReadOnlyDictionary<string, object?> updates, CancellationToken cancellationToken = default)
    {
        if (updates.Count == 0) return true;
        var payload = new Dictionary<string, object?>();
        var customFields = new Dictionary<string, object?>();
        foreach (var kv in updates)
        {
            if (kv.Key == "os_version" || kv.Key == "windows_feature_update")
                customFields[kv.Key] = kv.Value;
            else
                payload[kv.Key] = kv.Value;
        }
        if (customFields.Count > 0)
            payload["custom_fields"] = customFields;
        var payloadJson = JsonSerializer.Serialize(payload);
        var response = await SendWithRetryAsync(HttpMethod.Patch, $"{_baseUrl}/api/v1/hardware/{assetId}", () => new StringContent(payloadJson, Encoding.UTF8, "application/json"), cancellationToken).ConfigureAwait(false);
        return response.IsSuccessStatusCode;
    }

    public async Task<IReadOnlyList<SnipeItLookup>> GetModelsAsync(CancellationToken cancellationToken = default)
        => await FetchLookupsAsync("/api/v1/models", cancellationToken).ConfigureAwait(false);

    public async Task<IReadOnlyList<SnipeItLookup>> GetCategoriesAsync(CancellationToken cancellationToken = default)
        => await FetchLookupsAsync("/api/v1/categories", cancellationToken).ConfigureAwait(false);

    public async Task<IReadOnlyList<SnipeItLookup>> GetUsersAsync(CancellationToken cancellationToken = default)
    {
        var list = new List<SnipeItLookup>();
        var offset = 0;
        const int limit = 500;
        while (true)
        {
            var response = await SendWithRetryAsync(HttpMethod.Get, $"{_baseUrl}/api/v1/users?limit={limit}&offset={offset}", null, cancellationToken).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode) break;
            var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            var rows = root.ValueKind == JsonValueKind.Object && root.TryGetProperty("rows", out var r) ? r : root;
            if (rows.ValueKind != JsonValueKind.Array) break;
            int count = 0;
            foreach (var u in rows.EnumerateArray())
            {
                if (u.ValueKind != JsonValueKind.Object) continue;
                count++;
                var id = u.TryGetProperty("id", out var idEl) ? idEl.GetInt32() : 0;
                var first = u.TryGetProperty("first_name", out var fn) ? fn.GetString() ?? "" : "";
                var last = u.TryGetProperty("last_name", out var ln) ? ln.GetString() ?? "" : "";
                var uname = u.TryGetProperty("username", out var un) ? un.GetString() ?? "" : "";
                var name = !string.IsNullOrEmpty(first) || !string.IsNullOrEmpty(last) ? $"{first} {last}".Trim() : uname;
                list.Add(new SnipeItLookup { Id = id, Name = name });
            }
            if (count < limit) break;
            offset += count;
        }
        return list;
    }

    private async Task<IReadOnlyList<SnipeItLookup>> FetchLookupsAsync(string endpoint, CancellationToken cancellationToken)
    {
        var list = new List<SnipeItLookup>();
        var offset = 0;
        const int limit = 500;
        while (true)
        {
            var response = await SendWithRetryAsync(HttpMethod.Get, $"{_baseUrl}{endpoint}?limit={limit}&offset={offset}", null, cancellationToken).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode) break;
            var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            var rows = root.ValueKind == JsonValueKind.Object && root.TryGetProperty("rows", out var r) ? r : root;
            if (rows.ValueKind != JsonValueKind.Array) break;
            int count = 0;
            foreach (var item in rows.EnumerateArray())
            {
                if (item.ValueKind != JsonValueKind.Object) continue;
                count++;
                var id = item.TryGetProperty("id", out var idEl) ? idEl.GetInt32() : 0;
                var name = item.TryGetProperty("name", out var n) ? n.GetString() ?? "" : "";
                list.Add(new SnipeItLookup { Id = id, Name = name });
            }
            if (count < limit) break;
            offset += count;
        }
        return list;
    }

    private async Task<HttpResponseMessage> SendWithRetryAsync(HttpMethod method, string url, Func<HttpContent?>? bodyFactory, CancellationToken cancellationToken)
    {
        for (int attempt = 0; attempt <= MaxRetries; attempt++)
        {
            var client = CreateClient();
            var request = new HttpRequestMessage(method, url) { Content = bodyFactory?.Invoke() };
            var response = await client.SendAsync(request, cancellationToken).ConfigureAwait(false);

            if (response.StatusCode != HttpStatusCode.TooManyRequests || attempt == MaxRetries)
                return response;

            // Honor Retry-After header if present, otherwise exponential backoff
            var delay = response.Headers.RetryAfter?.Delta
                        ?? TimeSpan.FromSeconds(Math.Pow(2, attempt + 1));
            await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
        }
        throw new InvalidOperationException("Unreachable");
    }

    private HttpClient CreateClient()
    {
        var client = _httpClientFactory.CreateClient();
        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _getApiKey());
        client.DefaultRequestHeaders.Add("Accept", "application/json");
        return client;
    }
}

File: AssetSync-macOS/AssetSyncInfrastructure/Sources/AssetSyncInfrastructure/Api/SnipeItService.swift
import Foundation
import AssetSyncCore

/// Snipe-IT REST API client with rate-limit retry.
public final class SnipeItService: SnipeItServiceProtocol, @unchecked Sendable {
    private static let maxRetries = 3
    private let baseURL: String
    private let getApiKey: @Sendable () -> String
    private let session: URLSession

    public init(baseURL: String, getApiKey: @escaping @Sendable () -> String, session: URLSession = .shared) {
        self.baseURL = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        self.getApiKey = getApiKey
        self.session = session
    }

    public func getAssetBySerial(_ normalizedSerial: String) async throws -> Device? {
        let list = try await searchAssetsBySerial(normalizedSerial)
        return list.count == 1 ? list[0] : nil
    }

    public func searchAssetsBySerial(_ searchTerm: String) async throws -> [Device] {
        let encoded = searchTerm.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? searchTerm
        let (data, response) = try await sendWithRetry(.get, path: "/api/v1/hardware/byserial/\(encoded)")

        if (response as? HTTPURLResponse)?.statusCode == 404 { return [] }
        guard (response as? HTTPURLResponse)?.statusCode.isSuccessful == true else { return [] }

        let json = try JSONSerialization.jsonObject(with: data)

        if let arr = json as? [[String: Any]] {
            return arr.compactMap(parseDevice)
        }
        if let obj = json as? [String: Any] {
            if let rows = obj["rows"] as? [[String: Any]] {
                return rows.compactMap(parseDevice)
            }
            if obj["id"] != nil {
                return [parseDevice(obj)].compactMap { $0 }
            }
        }
        return []
    }

    public func createAsset(_ device: Device) async throws -> Device? {
        var payload: [String: Any] = [
            "name": device.deviceName ?? device.serialNumber ?? "",
            "serial": device.serialNumber ?? "",
            "model_id": device.snipeItModelId ?? 0,
            "status_id": 1,
        ]
        if let userId = device.snipeItAssignedUserId { payload["assigned_to"] = userId }
        if let catId = device.snipeItCategoryId { payload["category_id"] = catId }
        let tag = device.mdmAssetTag?.isEmpty == false
            ? device.mdmAssetTag!
            : "TEMP\(UUID().uuidString.prefix(8).uppercased())"
        payload["asset_tag"] = tag

        let body = try JSONSerialization.data(withJSONObject: payload)
        let (data, _) = try await sendWithRetry(.post, path: "/api/v1/hardware", body: body)

        guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }

        if let status = obj["status"] as? String, status == "error" {
            let messages = obj["messages"].map { "\($0)" } ?? "Unknown error"
            throw AssetSyncError.snipeItRejected(messages)
        }

        let created = (obj["payload"] as? [String: Any]) ?? obj
        var result = device
        result.snipeItAssetId = created["id"] as? Int
        return result
    }

    public func updateAsset(assetId: Int, updates: [String: Any]) async throws -> Bool {
        guard !updates.isEmpty else { return true }

        var payload: [String: Any] = [:]
        var customFields: [String: Any] = [:]
        for (key, value) in updates {
            if key == "os_version" || key == "windows_feature_update" {
                customFields[key] = value
            } else {
                payload[key] = value
            }
        }
        if !customFields.isEmpty { payload["custom_fields"] = customFields }

        let body = try JSONSerialization.data(withJSONObject: payload)
        let (_, response) = try await sendWithRetry(.patch, path: "/api/v1/hardware/\(assetId)", body: body)
        return (response as? HTTPURLResponse)?.statusCode.isSuccessful == true
    }

    public func getModels() async throws -> [SnipeItLookup] {
        try await fetchLookups(endpoint: "/api/v1/models")
    }

    public func getCategories() async throws -> [SnipeItLookup] {
        try await fetchLookups(endpoint: "/api/v1/categories")
    }

    public func getUsers() async throws -> [SnipeItLookup] {
        var list: [SnipeItLookup] = []
        var offset = 0
        let limit = 500
        while true {
            let (data, response) = try await sendWithRetry(.get, path: "/api/v1/users?limit=\(limit)&offset=\(offset)")
            guard (response as? HTTPURLResponse)?.statusCode.isSuccessful == true else { break }
            guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let rows = obj["rows"] as? [[String: Any]] else { break }
            for u in rows {
                let id = u["id"] as? Int ?? 0
                let first = u["first_name"] as? String ?? ""
                let last = u["last_name"] as? String ?? ""
                let uname = u["username"] as? String ?? ""
                let name = (!first.isEmpty || !last.isEmpty) ? "\(first) \(last)".trimmingCharacters(in: .whitespaces) : uname
                list.append(SnipeItLookup(id: id, name: name))
            }
            if rows.count < limit { break }
            offset += rows.count
        }
        return list
    }

    // MARK: - Private

    private enum HTTPMethodString: String { case get = "GET", post = "POST", patch = "PATCH" }

    private func sendWithRetry(_ method: HTTPMethodString, path: String, body: Data? = nil) async throws -> (Data, URLResponse) {
        for attempt in 0...Self.maxRetries {
            var request = URLRequest(url: URL(string: "\(baseURL)\(path)")!)
            request.httpMethod = method.rawValue
            request.setValue("Bearer \(getApiKey())", forHTTPHeaderField: "Authorization")
            request.setValue("application/json", forHTTPHeaderField: "Accept")
            if let body {
                request.httpBody = body
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            }
            let (data, response) = try await session.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            if status != 429 || attempt == Self.maxRetries {
                return (data, response)
            }
            let retryAfter = (response as? HTTPURLResponse)?
                .value(forHTTPHeaderField: "Retry-After")
                .flatMap(Double.init) ?? pow(2, Double(attempt + 1))
            try await Task.sleep(for: .seconds(retryAfter))
        }
        fatalError("Unreachable")
    }

    private func fetchLookups(endpoint: String) async throws -> [SnipeItLookup] {
        var list: [SnipeItLookup] = []
        var offset = 0
        let limit = 500
        while true {
            let (data, response) = try await sendWithRetry(.get, path: "\(endpoint)?limit=\(limit)&offset=\(offset)")
            guard (response as? HTTPURLResponse)?.statusCode.isSuccessful == true else { break }
            guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let rows = obj["rows"] as? [[String: Any]] else { break }
            for item in rows {
                let id = item["id"] as? Int ?? 0
                let name = item["name"] as? String ?? ""
                list.append(SnipeItLookup(id: id, name: name))
            }
            if rows.count < limit { break }
            offset += rows.count
        }
        return list
    }

    private func parseDevice(_ r: [String: Any]) -> Device? {
        guard r["id"] != nil else { return nil }
        var d = Device()
        d.snipeItAssetId = r["id"] as? Int
        d.serialNumber = r["serial"] as? String
        d.normalizedSerial = SerialNumberNormalizer.normalize(d.serialNumber)
        d.deviceName = (r["name"] as? String)?.removingHTMLEntities
        d.snipeItModelId = (r["model"] as? [String: Any])?["id"] as? Int
        d.snipeItAssignedUserId = (r["assigned_to"] as? [String: Any])?["id"] as? Int
        d.snipeItAssetTag = (r["asset_tag"] as? String)?.removingHTMLEntities
        return d
    }
}

public enum AssetSyncError: LocalizedError {
    case snipeItRejected(String)

    public var errorDescription: String? {
        switch self {
        case .snipeItRejected(let msg): return "Snipe-IT create rejected: \(msg)"
        }
    }
}

private extension Int {
    var isSuccessful: Bool { (200..<300).contains(self) }
}

private extension String {
    /// Basic HTML entity decoding for Snipe-IT responses.
    var removingHTMLEntities: String {
        replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
    }
}
